clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/cache/CatsOnDemandCacheUpdater.groovy

```diff
@@ -57,13 +61,27 @@ class CatsOnDemandCacheUpdater implements OnDemandCacheUpdater {
   OnDemandCacheUpdater.OnDemandCacheStatus handle(OnDemandAgent.OnDemandType type, Collection<OnDemandAgent> onDemandAgents, Map<String, ? extends Object> data) {
     boolean hasOnDemandResults = false
     for (OnDemandAgent agent : onDemandAgents) {
+      boolean atomicScheduler = false;
+      AgentScheduler scheduler = null
+      if (agent instanceof AgentSchedulerAware) {
+        scheduler = agent.agentScheduler
+        atomicScheduler = scheduler.atomic
+      }
       try {
+        AgentLock lock = null;
+        if (atomicScheduler && !(lock = scheduler.tryLock((Agent) agent))) {
+          hasOnDemandResults = true // force Orca to retry
+          continue;
+        }
         final long startTime = System.nanoTime()
         def providerCache = catsModule.getProviderRegistry().getProviderCache(agent.providerName)
         OnDemandAgent.OnDemandResult result = agent.handle(providerCache, data)
         if (result) {
+          if (atomicScheduler && !(scheduler.tryRelease(lock))) {
+            hasOnDemandResults = true // force Orca to retry
+          }
           if (result.cacheResult) {
-            hasOnDemandResults = !(result.cacheResult.cacheResults ?: [:]).values().flatten().isEmpty()
+            hasOnDemandResults = !(result.cacheResult.cacheResults ?: [:]).values().flatten().isEmpty() && !atomicScheduler
             agent.metricsSupport.cacheWrite {
```

Comment:
Yeah, I thought about this, and figured that since the since the `.handle()` call is on the order of seconds, while the time between the release and the write is on the order of a few instructions, this shouldn't be a problem (most of the time). If we release after write we miss out on fixing the race condition where `handle()` times out and writes old data. I thank for that reason, this current approach is safer. If we hit the current race condition, we are writing data captured from the same source within less than a second of each other. If we release after the write, we risk writing data that was captured after a timeout occurred, meaning minute-old data could overwrite fresh data.

Ideally we'd need an extra lock just around the write for this to be really atomic, perhaps on a per-resource basis even, but that's a lot of transactions. 

WDYT?